Adding disk and network benchmarks to redpanda self tests

graphcareful opened the pull request:
This pull request adds two benchmarks to the redpanda self-test framework, one for disk and another for network. These benchmarks replace the stub ones that were placed in the tree within `cluster/self_test_benchmarks.h`.

## Release Notes

### Features

  * Introduces new self-tests for redpanda allowing users to run mini-benchmarks on disk and network hardware to verify expected performance

jcsp commented:
Observations from a quick test of this on my workstation:
- The disk test works well, pushes up to ~5GB/s no problem with default settings.
- The network test had some unexpected behaviors:
  - It measured only ~300MB/s bandwidth between nodes running as processes on the same node (i.e. localhost networking)
  - Each time I ran it, I got ten `rpc - transport.cc:172 - Request timeout ...` warnings in the log, from one node (not always the same one).
  - In spite of those RPC timeouts, there isn't anything in the status that indicates an error happened while running the test, we should probably have an error counter.

jcsp commented on src/v/cluster/self_test_rpc_types.h:
```diff
@@ -39,33 +43,92 @@ inline std::ostream& operator<<(std::ostream& o, self_test_status sts) {
+            opts.name = obj["name"].GetString();
+        }
+        if (obj.HasMember("dsync")) {
+            opts.dsync = obj["dsync"].GetBool();
```
This parsing is faulty, assigning everything to `dsync`.  Would benefit from a unit test.

jcsp commented on src/v/cluster/self_test_rpc_types.h:
```diff
@@ -39,33 +43,92 @@ inline std::ostream& operator<<(std::ostream& o, self_test_status sts) {
+    /// Where files this benchmark will read/write to exist
+    std::filesystem::path dir{config::node().disk_benchmark_path()};
+    /// Open the file with O_DSYNC flag option
+    bool dsync{false};
```
Let's make dsync true by default: that is how Redpanda will behave.

jcsp commented on src/v/cluster/self_test/netcheck.cc:
```diff
@@ -0,0 +1,166 @@
+    /// the maximum amount of time for an individual RPC never exceeds the
+    /// remaining time of the test
+    auto now = ss::lowres_clock::now();
+    while (now < fiber_state.test_end() && !_cancelled) {
```
`now` is only used in this loop condition, so could just be inline `ss::lowres_clock::now() < ...`

jcsp commented on src/v/cluster/self_test/netcheck.cc:
```diff
@@ -0,0 +1,166 @@
+                _self,
+                ss::this_shard_id(),
+                peer,
+                10s,
```
I wonder if we should set the timeout to something lower, if we want to quickly detect that we've overwhelmed a peer.  Or, if we would like to permit a big queue of RPCs, then maybe the timeout should be the test duration.

jcsp commented on src/v/cluster/self_test/netcheck.cc:
```diff
@@ -0,0 +1,166 @@
+  run_fiber_opts& fiber_state) {
+    if (!reply) {
+        if (reply.error() == rpc::errc::client_request_timeout) {
+            /// timeout expired, let above loop continue
```
We should count these timeouts and include it in the result

jcsp commented on src/v/cluster/self_test/netcheck.cc:
```diff
@@ -0,0 +1,166 @@
+        /// will be thrown away
+        if (fiber_state.not_started()) {
+            /// still waiting for server to be available
+            co_await ss::sleep(200ms);
```
Non-abortable sleeps are undesirable (even when short), is there an abort source we could pass in?

jcsp reviewed (commented):
This is pretty cool!  I got a kick out of running it on my workstation and seeing the disk light up.

I've left a bunch of comments, but nothing that makes me worry too much about getting this polished off by end of next week.

graphcareful commented on src/v/cluster/self_test/diskcheck.cc:
```diff
@@ -0,0 +1,212 @@
+ss::future<std::vector<self_test_result>> diskcheck::run(diskcheck_opts opts) {
+    vassert(_gate.get_count() == 0, "Benchmark already in progress");
+    gate_guard g{_gate};
+    co_await ss::futurize_invoke(validate_options, opts);
```
The idea was for the function to throw seastar exceptional futures and not a regular c++ future. Didn't want to make that method `validate_options` a coroutine if there was no async work.

jcsp replied:
> The idea was for the function to throw seastar exceptional futures and not a regular c++ future

I confess I've never delved into how this works, but I would have thought that the type of the future would be determined at the point we call the function?  So if we call a synchronous function from a coroutine like diskcheck::run that returns a ss::future, wouldn't the exception get turned into a seastar exceptional future without any extra work on our part?

andrwng replied:
I agree it seems like a reasonable expectation but it looks like that isn't the case, looking at the example in 7.1 https://docs.seastar.io/master/split/7.html

dotnwat replied:
>The idea was for the function to throw seastar exceptional futures and not a regular c++ future. Didn't want to make that method validate_options a coroutine if there was no async work.

wouldn't the co_await then just unwrap and throw c++ exception anyway?

graphcareful replied:
within the context of a coroutine it would make senese to me that after unwrap and throw, it would throw an exceptional future though

dotnwat replied:
in a coroutine if you `throw X` it'll turn into automatically an exceptional future holding X. you can't return an exceptional future (or any other future) in a coroutine--just values.

graphcareful commented on src/v/cluster/self_test/diskcheck.cc:
```diff
@@ -0,0 +1,212 @@
+    _cancelled = false;
+    _opts = opts;
+    _last_pos = 0;
+    const auto fname = ssx::sformat(
```
Nice feedback, going to integrate with `node::local_monitor`

graphcareful commented on src/v/cluster/self_test/diskcheck.cc:
```diff
@@ -0,0 +1,212 @@
+        }
+        r.push_back(std::move(result));
+    }
+    if (!_opts.skip_read) {
```
I'm actually not sure. If this is the case i'm fine removing the `skip_write` option, in either case as it stands the plan would not be to set either of these to false with any of the default options or any options I will expose.

graphcareful replied:
Current idea is to leave this in, but via rpk never expose the `skip_write` option to the user

graphcareful commented on src/v/cluster/self_test/diskcheck.cc:
```diff
@@ -0,0 +1,212 @@
+uint64_t diskcheck::get_pos() {
+    uint64_t pos = _last_pos;
+    _last_pos += _opts.request_size;
+    if (_last_pos >= _opts.file_size()) {
```
I think that is ok, but the return value seems sus, its like returning the current value, not the current value incremented by request size

graphcareful replied:
I think this is OK, `_last_pos` is incremented by the request size before the conditional is checked

graphcareful commented:
Changes in recent force-pushes:
- Implemented all of Johns feedback above with the exception of the outstanding question and debugging the connection timeouts within the network benchmark. Will tackle those next.

graphcareful commented:
Changes in recent [force-push](https://github.com/redpanda-data/redpanda/compare/fdba86a8725ce29060a0c91fc71f0a061bb9f621..9bd4389f26fb7132827848b187a581f7fda4ab3f) 
- Rebased against dev

graphcareful commented:
Changes since last [force-push](https://github.com/redpanda-data/redpanda/compare/9bd4389f26fb7132827848b187a581f7fda4ab3f..8a85f393a2ba2e38f70842223db421d1784ac6b6)
- Cosmetic change to exception class names
- Method to break network requests into fragments of a greater iobuf when request sizes are larger then a few kb
- Generating random request data blobs instead of requests with zeroed out data
- Validation for network tests
- Reworked some of the logic in `run_fiber_state` for better handling of proper start/stop of tests
- Having the network benchmark server not return the same data as was sent to it. I've looked at a few other network benchmarks, they do not do this.
- Fixed a bug where network benchmark not stop if you passed in a list of custom node ids to run the network tests on.

Note: For those observing slower then normal benchmark results for the network test, I've tracked that down to running redpanda in debug vs release mode. In release mode i'm seeing results that are magnitudes greater then in debug mode for the network benchmark, example:
```   
NAME        Network Test 8192b packet size
INFO        Test performed against node: 2
TEST ID:    99d96d01-d534-40cb-9062-8f7e1c50ac0a
TIMEOUTS:   0
DURATION    5000ms
IOPS        12388 requests/sec
THROUGHPUT  96.78MiB/sec
LATENCY     P50    P90     P99     P999    Max
            767us  1023us  1599us  1855us  7167us

NAME        Network Test 8192b packet size
INFO        Test performed against node: 1
TEST ID:    99d96d01-d534-40cb-9062-8f7e1c50ac0a
TIMEOUTS:   0
DURATION    5000ms
IOPS        12388 requests/sec
THROUGHPUT  96.22MiB/sec
LATENCY     P50    P90     P99     P999    Max
            895us  1343us  2175us  3071us  4351us
```
For comparison, in debug mode I had observed the iops rate at lower then 400 rps.

TODO: Fix some of the rpc timeouts that appear in the logs

andrwng reviewed (commented):
Structurally looks pretty good

graphcareful commented on src/v/cluster/self_test_backend.cc:
```diff
@@ -41,7 +41,12 @@ ss::future<std::vector<self_test_result>> self_test_backend::do_start_test(
     for (auto& dto : dtos) {
         try {
-            vlog(clusterlog.info, "Starting disk self test...");
+            if (std::filesystem::exists(dto.dir)) {
+                /// Ensure no leftover large files in the event there was a
+                /// crash mid run and cleanup didn't get a chance to occur
+                std::filesystem::remove_all(dto.dir);
+            }
+            std::filesystem::create_directory(dto.dir);
```
I thought so too, but that prevents diskcheck from being run concurrently. Not that it does now, but thats what i had in mind for this

graphcareful replied:
Taking this feedback

graphcareful commented on src/v/cluster/self_test/diskcheck.cc:
```diff
@@ -0,0 +1,242 @@
+ss::future<metrics> diskcheck::do_run_benchmark(ss::file& file) {
+    auto irange = boost::irange<uint16_t>(0, _opts.parallelism);
+    auto start = ss::lowres_clock::now();
+    static const auto five_seconds_us = 500000;
```
This constant is passed to the hdr_hist constructor. It should be == to the largest value you'd expect to ever put in your histogram.

graphcareful commented on src/v/cluster/self_test/netcheck.cc:
```diff
@@ -0,0 +1,213 @@
+        fzs.push_back(i);
+    }
+    auto buffers = co_await ssx::parallel_transform(fzs, [](size_t size) {
+        ss::temporary_buffer<char> buf(size);
+        random_generators::fill_buffer_randomchars(buf.get_write(), size);
```
the input parameter `sz` is divided up into chunks of `8192` pieces (fragments). `size` is each subsequent fragments size. This does no bounds checking (other then the limit of `size_t` input type itself) because bounds checking was already performed by `validate_options`. So `max` fragment size is 8192 bytes (it would take 15625 fragments to == 128MB) , which would comprise of 1 `iobuf` for 1 `netcheck_request`

graphcareful commented on src/v/cluster/self_test/diskcheck.h:
```diff
@@ -0,0 +1,126 @@
+    ss::future<> run_benchmark_fiber(
+      ss::lowres_clock::time_point start, ss::file& file, metrics& m);
+
+    uint64_t get_pos();
```
all fibers actually share this `_pos` offset. Adding descriptive comment

andrwng commented on src/v/cluster/self_test/tests/self_test_request_test.cc:
```diff
@@ -0,0 +1,46 @@
+    req = cluster::make_netcheck_request(model::node_id{0}, size).get0().buf;
+    BOOST_CHECK_EQUAL(req.size_bytes(), size);
+    BOOST_CHECK_EQUAL(std::distance(req.cbegin(), req.cend()), 53);
+    BOOST_CHECK(check_frags(req.cbegin(), (--(--req.cend()))));
```
Why the double --? Doesn't `check_frags()` skip comparing `end`?

graphcareful replied:
No, but I think modifying `check_frags()` to make it more obvious to whats going on is the move

andrwng commented on src/v/cluster/self_test/tests/self_test_request_test.cc:
```diff
@@ -0,0 +1,46 @@
+          return true;
+      };
+
+    auto req = cluster::make_netcheck_request(model::node_id{0}, 52).get0().buf;
+    BOOST_CHECK_EQUAL(req.size_bytes(), 52);
+    BOOST_CHECK_EQUAL(std::distance(req.cbegin(), req.cend()), 1);
+    BOOST_CHECK_EQUAL(req.cbegin()->size(), 52);
```
Maybe add a high level comment describing what this is doing? E.g. "Check that all buffers have size 8192 except the last, which fills in any remaining space." or somesuch

andrwng reviewed (commented):
Looking good, just some test suggestions

The new DT test doesn't look happy yet either

jcsp commented:
I think this is the PR where this feature is really usable, so please can you add the release notes for the overall self test feature?

jcsp commented on src/v/cluster/self_test/netcheck.cc:
```diff
@@ -0,0 +1,186 @@
+        throw;
+    } catch (const std::exception& ex) {
+        vlog(clusterlog.error, "Unforseen exception encountered: {}", ex);
+        throw netcheck_exception(ex.what());
```
In general, when we see an unexpected exception type, it's better to let it bubble up to a higher level rather than changing the type and rethrowing -- e.g. in this case do_start_test already knows how to handle a std::exception and include its .what() in the results, so this branch isn't needed.

jcsp commented on src/v/cluster/self_test_backend.cc:
```diff
@@ -79,19 +94,18 @@ get_status_response self_test_backend::start_test(start_test_request req) {
         vlog(
-          clusterlog.info, "Starting redpanda self-tests with id: {}", req.id);
+          clusterlog.debug, "Request to start self-tests with id: {}", req.id);
         (void)do_start_test(req.dtos, req.ntos)
```
This line wasn't added in this PR, but rather than using `(void)` to background a coroutine that includes a gate holder, use ssx::spawn_with_gate

jcsp commented on src/v/cluster/self_test/netcheck.cc:
```diff
@@ -0,0 +1,186 @@
+          reply.error() == rpc::errc::client_request_timeout
+          || reply.error() == rpc::errc::exponential_backoff) {
+            co_await ss::sleep_abortable(200ms, _as);
+            throw omit_metrics_measurement_exception();
```
This is just skipping the measurement, right?  We should count the failures and include an error count in the result object.

graphcareful replied:
Nice catch, something got lost in merge

jcsp commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -125,6 +125,37 @@ ss::future<> self_test_frontend::start() { return ss::now(); }
+            if (requested_id != node_id) {
+                vlog(
+                  clusterlog.debug,
+                  "Adding {client: {} server: {}} pair to netcheck plan",
```
Format string has more placeholders than vlog has arguments

andrwng reviewed (approved):
LGTM! Nice work

graphcareful commented on tests/rptest/tests/self_test_test.py:
```diff
@@ -146,7 +167,7 @@ def test_self_test_cancellable(self):
-                assert 'error' not in report
+                assert 'error' not in report, report['error']
                 assert 'warning' in report
                 assert 'duration' in report
```
It was technically incorrect, if the cancel report comes before any test could have possibly stopped, then all tests should return as cancelled

graphcareful commented on src/v/cluster/self_test_backend.cc:
```diff
@@ -74,24 +100,25 @@ ss::future<std::vector<self_test_result>> self_test_backend::do_start_test(
 }
 
 get_status_response self_test_backend::start_test(start_test_request req) {
-    auto gate_holder = _gate.hold();
```
Theres gate holding here but now its done by `ssx::spawn_with_gate`. The background work is the async work in this context

rystsov commented on tests/rptest/tests/self_test_test.py:
```diff
@@ -99,10 +122,14 @@ def test_self_test_node_crash(self):
-                assert report['duration'] >= 1000, report['duration']
+                # Errors related to crashed node are allowed, for example a network test on a good node that had attempted to connect to the crashed node
+                if 'error' in report:
+                    report[
```
missed assert?

graphcareful replied:
This is a test that crashes a node so a failure can be expected

graphcareful commented:
Change in [force-push](https://github.com/redpanda-data/redpanda/compare/fdbf58815ad318297a50e5e1ab734e0a40c4a7a0..80e370ce063938a3817e63818da00960380a0a63) 
- Fix logging issue that cause CI issue when rp running in debug mode

graphcareful commented:
/ci-repeat 2 dt-repeat=25 skip-unit tests/rptest/tests/self_test_test.py
